File: AWS/iac-microbenchmark/ec2/typescript/ecc256_verify.ts
#!/usr/bin/env node
import { KMSClient, VerifyCommand } from '@aws-sdk/client-kms';
import * as process from 'process';
import * as base64 from 'base64-js';

// Initialize the AWS KMS client with the specified region
const kmsClient = new KMSClient({
  region: 'us-east-1' // Specify the desired region
});

async function main(): Promise<void> {
  try {
    // Get the JSON input from command line arguments
    const requestJsonRaw: string = process.argv[2];

    // Parse the JSON input
    const requestJson = JSON.parse(requestJsonRaw);

    // Extract the message and signature from the parsed JSON
    const message: string = requestJson.get('message');
    const signatureB64: string = requestJson.get('signature');

    // Decode the base64-encoded signature
    const signature = base64.toByteArray(signatureB64);

    // Retrieve the KMS key ARN from environment variables
    const eccKmsKeyId = process.env['ECC256_KMS_KEY_ARN'];

    if (!eccKmsKeyId) {
      throw new Error("ECC256_KMS_KEY_ARN environment variable is not set");
    }

    // Verify the signature using KMS
    const verifyCommand = new VerifyCommand({
      KeyId: eccKmsKeyId,
      Message: Buffer.from(message, 'utf-8'),
      MessageType: 'RAW',
      Signature: signature,
      SigningAlgorithm: 'ECDSA_SHA_256' // Use 'ECDSA_SHA_384' for P-384
    });

    const response = await kmsClient.send(verifyCommand);

    // Output the result of the verification
    console.log(JSON.stringify({ verified: response.SignatureValid }));
  } catch (error) {
    // Handle and log any errors that occur
    console.error("Error:", error);
  }
}

// Execute the main function
main();

File: AWS/iac-microbenchmark/ec2/python/ecc256_verify.py
import json
import boto3
import base64
from botocore.exceptions import ClientError
import os
import sys
import json

kms_client = boto3.client('kms')

def main():
    # Get the name argument from sys.argv
    request_json_raw = sys.argv[1]

    request_json = json.loads(request_json_raw)
    # Extract the message from the event payload
    message = request_json.get('message')

    signature_b64 = request_json.get('signature')

    signature = base64.b64decode(signature_b64)

    ecc_kms_key_id = os.environ['ECC256_KMS_KEY_ARN']
     # Verify the signature
    response = kms_client.verify(
        KeyId=ecc_kms_key_id,
        Message=message.encode('utf-8'),
        MessageType='RAW',
        Signature=signature,
        SigningAlgorithm='ECDSA_SHA_256'  # Use 'ECDSA_SHA_384' for P-384
    )
        
    print(json.dumps({'verified': response['SignatureValid']}))

if __name__ == "__main__":
    main()
